### PR TITLE
added kadane algorithm function

### PR DESCRIPTION
**Pull request by ayush523**
<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1079"></a>

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1079"></a>

### CODE REVIEW

**Comment by ayush523** [comment]
What is the problem with the file that I have added?Please help

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please read the [Contribution Guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md) and fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/pull/1079/checks?check_run_id=1119386114) warnings.

**Comment by Panquesito7** [comment]
The file does not need to be in a specific folder for this algorithm.
It can be in the `others` folder, or in its appropriate folder. 🙂

**Comment by ayush523** [comment]
Is that why the tests are failing?

**Comment by ayush523** [comment]
I have made the changes.Please check.

File: kadane_algorithm/kadane.cpp [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    for (int i = 0; i < n.size(); i++) {
+        std::cout << "Enter value of n[" << i << "]"
+                  << "\n";
+        std::cin >> n[i];
+    }
+    int max_sum = maxSubArray(n);
+    std::cout << "Maximum subarray sum is " << max_sum;
+
+    return 0;
+}
```
Comment (Panquesito7) [inline kadane_algorithm/kadane.cpp:28]:
```suggestion
}

```

File: kadane_algorithm/kadane.cpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    int curr_sum = 0;
+    int max_sum = INT_MIN;
+    for (int i = 0; i < n.size(); i++) {
+        curr_sum += n[i];
+        max_sum = std::max(max_sum, curr_sum);
+        curr_sum = std::max(curr_sum, 0);
+    }
+    return max_sum;
+}
+int main() {
```
Comment (Panquesito7) [inline kadane_algorithm/kadane.cpp:15]:
```suggestion

/**
 * @brief Main function
 * @returns 0 on exit
 */
int main() {
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Looks much better, good work. 😄 👍
Now, the next thing we shall do is to add documentation.

Confused? See #962 as a reference. 🙂

**Comment by ayush523** [comment]
Thanx. I did the changes which you suggested.

File: kadane_algorithm/kadane.cpp [hunk 3]
```diff
@@ -0,0 +1,33 @@
+#include <array>
+#include <climits>
+#include <iostream>
+
+int maxSubArray(const std::array<int, 5> &n) {
+    int curr_sum = 0;
+    int max_sum = INT_MIN;
+    for (int i = 0; i < n.size(); i++) {
```
Comment (Panquesito7) [inline kadane_algorithm/kadane.cpp:8]:
```suggestion
    for (int i : n) {
```

File: kadane_algorithm/kadane.cpp [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    }
+    return max_sum;
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    std::array<int, 5> n;
```
Comment (Panquesito7) [inline kadane_algorithm/kadane.cpp:21]:
```suggestion
    std::array<int, 5> n {};
```

**Comment by ayush523** [comment]
Any more changes required?

**Comment by Panquesito7** [comment]
Yes, please add documentation like in #962.

File: kadane_algorithm/kadane.cpp [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    for (int i = 0; i < n.size(); i++) {
+        std::cout << "Enter value of n[" << i << "]"
+                  << "\n";
+        std::cin >> n[i];
+    }
+    int max_sum = maxSubArray(n);  // calling maxSubArray function
+    std::cout << "Maximum subarray sum is " << max_sum;  // Printing the answer
+
+    return 0;
+}
```
Comment (Panquesito7) [inline kadane_algorithm/kadane.cpp:34]:
```suggestion
}

```

File: kadane_algorithm/kadane.cpp [hunk 6]
```diff
@@ -0,0 +1,34 @@
+#include <array>
```
Comment (Panquesito7) [inline kadane_algorithm/kadane.cpp:1]: Provide more information
in the `@details` part of the block.

Confused? See #962 as a reference. 🙂
```suggestion
/**
 * @file
 * @brief Implementation of [Kadane Algorithm](https://en.wikipedia.org/wiki/Kadane%27s_algorithm)
 *
 * @details
 * <full-description here>
 *
 * @author [Ayush Singh](https://github.com/ayush523)
 */
#include <array>
```

File: kadane_algorithm/kadane.cpp [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+#include <climits>
+#include <iostream>
+
+int maxSubArray(const std::array<int, 5> &n) {
```
Comment (Panquesito7) [inline kadane_algorithm/kadane.cpp:5]: Just the documentation template.

Provide brief information about the
function, parameters and return statement.

```suggestion
/**
 * @brief description
 * @param n description
 * @returns description
 */
int maxSubArray(const std::array<int, 5> &n) {
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: kadane_algorithm/kadane.cpp [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+ * @file kadane.cpp
+ * @brief Implementation of [Kadane
+ * Algorithm] (https://en.wikipedia.org/wiki/Kadane%27s_algorithm)
+ *
+ * @details
+ * Kadane algorithm is used to find the maximum sum subarray in an array and
+ * maximum sum subarray problem is the task of finding a contiguous subarray
+ * with the largest sum
+ *
+ * Algorithm
```
Comment (Panquesito7) [inline kadane_algorithm/kadane.cpp:11]:
```suggestion
 * ### Algorithm
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: kadane_algorithm/kadane.cpp (hunk 7 above)
Comment (Panquesito7) [reply kadane_algorithm/kadane.cpp:29]: See #962 for how to document functions.

**Comment by kvedala** [comment]
Is this a repeat of #606 ?

**Comment by Panquesito7** [comment]
It seems it is, sorry, my bad!

What if in this PR we improve the current algorithm by adding
documentation and make it more a C++ program?

**Comment by ayush523** [comment]
The issue was to create kadane's algorithm.I did that on my own.I haven't copied anyone's code.

**Comment by ayush523** [comment]
@Panquesito7 -I have already added documentation which is missing in #606.

**Comment by kvedala** [comment]
> The issue was to create kadane's algorithm.I did that on my own.I haven't copied anyone's code.

@ayush523 Good to hear that the algorithm and code is your own.
Ensure that:
1. the contributed code and/or the algorithm are different implementations for the same problem
2. the filename should be changed to `dynamic_programming/kadane2.cpp`. make sure no new folders are added.
3. appropriate documentation to explain the new algorithm

Thank you

**Comment by kvedala** [comment]
> The issue was to create kadane's algorithm.I did that on my own.I haven't copied anyone's code.

**Comment by ayush523** [comment]
@kvedala -Sir I have done the required changes.Please check.

**Comment by kvedala** [comment]
> @kvedala -Sir I have done the required changes.Please check.

@Panquesito7 is the reviewer and he will continue the reviews.

**Comment by kvedala** [comment]
> It seems it is, sorry for not noticing!
> 
> What if in this PR we improve the current algorithm by adding
> documentation and make it more a C++ program?

Its ok, it is easy to miss the history :)

File: dynamic_programming/kadane2.cpp [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+/**
+ * @brief description
+ * @param n description
+ * @returns description
+ */
+/**
+ * @brief description
+ * @param n description
+ * @returns description
+ */
```
Comment (Panquesito7) [inline dynamic_programming/kadane2.cpp:33]: Provide a brief description about the parameters, return statement and description of the function.
```suggestion
/**
 * @brief description
 * @param n description
 * @returns description
 */
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: dynamic_programming/kadane2.cpp [hunk 10]
```diff
@@ -0,0 +1,58 @@
+/**
+ * @file kadane.cpp
```
Comment (Panquesito7) [inline dynamic_programming/kadane2.cpp:2]:
```suggestion
 * @file
```

File: dynamic_programming/kadane2.cpp [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+ * if it is greater than curr_sum
+ *
+ * @author [Ayush Singh](https://github.com/ayush523)
+ */
+#include <array>
+#include <climits>
+#include <iostream>
+
+/**
+ * @brief maxSubArray function
```
Comment (Panquesito7) [inline dynamic_programming/kadane2.cpp:25]: What does this function do?
Document it in the brief description.

If wanted, you can also add further information
about the function using `@details`.

File: dynamic_programming/kadane2.cpp [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+ * @param n array of values and size of the array
+ * @returns the value of maximum subarray sum
+ */
```
Comment (Panquesito7) [inline dynamic_programming/kadane2.cpp:28]:
```suggestion
 */
 template <size_t N>
```

File: dynamic_programming/kadane2.cpp [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+int maxSubArray(const std::array<int, 5> &n) {
```
Comment (Panquesito7) [inline dynamic_programming/kadane2.cpp:29]:
```suggestion
int maxSubArray(const std::array<int, N> &n) {
```

File: dynamic_programming/kadane2.cpp (hunk 11 above)
Comment (Panquesito7) [inline dynamic_programming/kadane2.cpp:25]:
```suggestion
 * @brief maxSubArray function
 * @tparam N number of array size
```

File: dynamic_programming/kadane2.cpp [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+int main() {
+    std::array<int, 5> n{};  // declaring array
+
+    // taking values of elements from user
+    for (int i = 0; i < n.size(); i++) {
+        std::cout << "Enter value of n[" << i << "]"
+                  << "\n";
+        std::cin >> n[i];
+    }
+    int max_sum = maxSubArray(n);  // calling maxSubArray function
```
Comment (Panquesito7) [inline dynamic_programming/kadane2.cpp:54]:
```suggestion
    int max_sum = dynamic_programming::maxSubArray<N>(n);  // calling maxSubArray function
```

File: dynamic_programming/kadane2.cpp [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+ * segment among all positive segments(curr_sum is used for this)
+ * Each time we get a positive sum we compare it with max_sum and update max_sum
+ * if it is greater than curr_sum
+ *
+ * @author [Ayush Singh](https://github.com/ayush523)
+ */
+#include <array>
+#include <climits>
+#include <iostream>
+
```
Comment (Panquesito7) [inline dynamic_programming/kadane2.cpp:23]:
```suggestion
/**
 * @namespace dynamic_programming
 * @brief Dynamic Programming algorithms
 */
namespace dynamic_programming {
```

File: dynamic_programming/kadane2.cpp [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+    int curr_sum =
+        0;  // declaring a variable named as curr_sum and initialized it to 0
+    int max_sum = INT_MIN;  // Initialized max_sum to INT_MIN
+    for (int i : n) {  // for loop to iterate over the elements of the array
+        curr_sum += n[i];
+        max_sum = std::max(max_sum, curr_sum);  // getting the maximum value
+        curr_sum = std::max(curr_sum, 0);  // updating the value of curr_sum
+    }
+    return max_sum;  // returning the value of max_sum
+}
```
Comment (Panquesito7) [inline dynamic_programming/kadane2.cpp:39]:
```suggestion
}
}  // namespace dynamic_programming
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Comment by ayush523** [comment]
@Panquesito7 -I don't understand why the code is not working after the commits suggested.Please help.

File: .vscode/c_cpp_properties.json [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+                "_UNICODE"
+            ],
+            "compilerPath": "C:\\TDM-GCC-64\\bin\\gcc.exe",
+            "cStandard": "gnu11",
+            "cppStandard": "c++17",
+            "intelliSenseMode": "gcc-x64"
+        }
+    ],
+    "version": 4
+}
```
Comment (Panquesito7) [inline .vscode/c_cpp_properties.json:20]: Please remove this file.

File: .vscode/settings.json [hunk 18]
```diff
@@ -2,5 +2,6 @@
      "C_Cpp.clang_format_style": "{ BasedOnStyle: Google, UseTab: Never, IndentWidth: 4, TabWidth: 4, AllowShortIfStatementsOnASingleLine: false, IndentCaseLabels: true, ColumnLimit: 80, AccessModifierOffset: -3, AlignConsecutiveMacros: true }",
      "editor.formatOnSave": true,
      "editor.formatOnType": true,
-     "editor.formatOnPaste": true
+     "editor.formatOnPaste": true,
+     "C_Cpp.errorSquiggles": "Disabled"
```
Comment (Panquesito7) [inline .vscode/settings.json:6]: Please undo this change.

File: dynamic_programming/kadane2.cpp [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+ */
+namespace dynamic_programming {
+/**
+ * @brief maxSubArray function is used to calculate the maximum sum subarray
+ * and returns the value of maximum sum which is stored in the variable max_sum
+ * @param n is the array n
+ * @returns the value of maximum subarray sum
+ */
+template <size_t N>
+
```
Comment (Panquesito7) [inline dynamic_programming/kadane2.cpp:35]:
```suggestion
```

File: dynamic_programming/kadane2.cpp [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+#include <iostream>
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming algorithms
+ */
+namespace dynamic_programming {
+/**
+ * @brief maxSubArray function is used to calculate the maximum sum subarray
+ * and returns the value of maximum sum which is stored in the variable max_sum
+ * @param n is the array n
```
Comment (Panquesito7) [inline dynamic_programming/kadane2.cpp:31]:
```suggestion
 * @param n array where numbers are saved
```

File: dynamic_programming/kadane2.cpp [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+    return max_sum;  // returning the value of max_sum
+}
+}  // namespace dynamic_programming
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    std::array<int, 5> n{};  // declaring array
```
Comment (Panquesito7) [inline dynamic_programming/kadane2.cpp:54]:
```suggestion
int main() {
    const int N = 5;
    std::array<int, N> n{};  // declaring array
```

File: dynamic_programming/kadane2.cpp [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+#include <climits>
+#include <iostream>
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming algorithms
+ */
+namespace dynamic_programming {
+/**
+ * @brief maxSubArray function is used to calculate the maximum sum subarray
+ * and returns the value of maximum sum which is stored in the variable max_sum
```
Comment (Panquesito7) [inline dynamic_programming/kadane2.cpp:30]:
```suggestion
 * and returns the value of maximum sum which is stored in the variable max_sum
 * @tparam N number of array size
```

File: dynamic_programming/kadane2.cpp [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+ * @author [Ayush Singh](https://github.com/ayush523)
+ */
+#include <array>
+#include <climits>
+#include <iostream>
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming algorithms
+ */
+namespace dynamic_programming {
```
Comment (Panquesito7) [inline dynamic_programming/kadane2.cpp:27]:
```suggestion
namespace dynamic_programming {
/**
 * @namespace kadane
 * @brief Functions for [Kadane Algorithm](https://en.wikipedia.org/wiki/Kadane%27s_algorithm)
 */
namespace kadane {
```

File: dynamic_programming/kadane2.cpp [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+        0;  // declaring a variable named as curr_sum and initialized it to 0
+    int max_sum = INT_MIN;  // Initialized max_sum to INT_MIN
+    for (int i : n) {  // for loop to iterate over the elements of the array
+        curr_sum += n[i];
+        max_sum = std::max(max_sum, curr_sum);  // getting the maximum value
+        curr_sum = std::max(curr_sum, 0);  // updating the value of curr_sum
+    }
+    return max_sum;  // returning the value of max_sum
+}
+}  // namespace dynamic_programming
```
Comment (Panquesito7) [inline dynamic_programming/kadane2.cpp:47]:
```suggestion
}  // namespace kadane
}  // namespace dynamic_programming
```

File: dynamic_programming/kadane2.cpp [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+    std::array<int, 5> n{};  // declaring array
+
+    // taking values of elements from user
+    for (int i = 0; i < n.size(); i++) {
+        std::cout << "Enter value of n[" << i << "]"
+                  << "\n";
+        std::cin >> n[i];
+    }
+    int max_sum =
+        dynamic_programming::maxSubArray<N>(n);  // calling maxSubArray function
```
Comment (Panquesito7) [inline dynamic_programming/kadane2.cpp:63]:
```suggestion
        dynamic_programming::kadane::maxSubArray<N>(n);  // calling maxSubArray function
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: .vscode/settings.json [hunk 26]
```diff
@@ -1,6 +1,6 @@
 {
-     "C_Cpp.clang_format_style": "{ BasedOnStyle: Google, UseTab: Never, IndentWidth: 4, TabWidth: 4, AllowShortIfStatementsOnASingleLine: false, IndentCaseLabels: true, ColumnLimit: 80, AccessModifierOffset: -3, AlignConsecutiveMacros: true }",
-     "editor.formatOnSave": true,
-     "editor.formatOnType": true,
-     "editor.formatOnPaste": true
+  "C_Cpp.clang_format_style": "{ BasedOnStyle: Google, UseTab: Never, IndentWidth: 4, TabWidth: 4, AllowShortIfStatementsOnASingleLine: false, IndentCaseLabels: true, ColumnLimit: 80, AccessModifierOffset: -3, AlignConsecutiveMacros: true }",
+  "editor.formatOnSave": true,
+  "editor.formatOnType": true,
+  "editor.formatOnPaste": true
```
Comment (Panquesito7) [inline .vscode/settings.json:5]:
```suggestion
     "C_Cpp.clang_format_style": "{ BasedOnStyle: Google, UseTab: Never, IndentWidth: 4, TabWidth: 4, AllowShortIfStatementsOnASingleLine: false, IndentCaseLabels: true, ColumnLimit: 80, AccessModifierOffset: -3, AlignConsecutiveMacros: true }",
     "editor.formatOnSave": true,
     "editor.formatOnType": true,
     "editor.formatOnPaste": true
```

File: dynamic_programming/kadane2.cpp [hunk 27]
```diff
@@ -0,0 +1,69 @@
...
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming algorithms
+ */
+namespace dynamic_programming {
+/**
+ * @brief maxSubArray function is used to calculate the maximum sum subarray
+ * and returns the value of maximum sum which is stored in the variable max_sum
+ * @param n array where numbers are saved
+ * @returns the value of maximum subarray sum
+ * @tparam N number of array size
```
Comment (Panquesito7) [inline dynamic_programming/kadane2.cpp:33]:
```suggestion
 * @brief maxSubArray function is used to calculate the maximum sum subarray
 * and returns the value of maximum sum which is stored in the variable max_sum
 * @tparam N number of array size
 * @param n array where numbers are saved
 * @returns the value of maximum subarray sum
```

File: dynamic_programming/kadane2.cpp [hunk 28]
```diff
@@ -0,0 +1,69 @@
...
+ */
+namespace kadane {
```
Comment (Panquesito7) [inline dynamic_programming/kadane2.cpp:35]:
```suggestion
```

File: dynamic_programming/kadane2.cpp [hunk 29]
```diff
@@ -0,0 +1,69 @@
...
+ * @author [Ayush Singh](https://github.com/ayush523)
+ */
+#include <array>
+#include <climits>
+#include <iostream>
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming algorithms
+ */
+namespace dynamic_programming {
```
Comment (Panquesito7) [inline dynamic_programming/kadane2.cpp:27]:
```suggestion
namespace dynamic_programming {
/**
 * @namespace kadane
 * @brief Functions for [Kadane](https://en.wikipedia.org/wiki/Kadane%27s_algorithm) algorithm.
 */
namespace kadane {
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Review by Panquesito7 — APPROVED** [review]
LGTM, good work. 👍 😄

**Comment by ayush523** [comment]
Thank you @Panquesito7

**Comment by kvedala** [comment]
@Panquesito7 do you confirm the algorithm is a different implementation than the existing one? thank you

**Comment by Panquesito7** [comment]
Yes, implementation looks different. 🙂
I like this one rather than the current one:

- Has very good documentation
- Uses C++ standard-libraries/functions
- Well-structured and as per the [repository standards](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md)

Can be merged. 👍

**Pull request merged**